### PR TITLE
[FLINK-24741]Deprecate FileRecordFormat

### PR DESCRIPTION
**Pull request by JingGe**
## What is the purpose of the change

The FileRecordFormat and StreamFormat have too much commons. This makes user confused. The main motivation for removing it is the inherent design flaw in the batching of FileRecordFormat: StreamFormat can guarantee that only a certain amount of memory is being used (unless a single record exceeds that already), but FileRecordFormat can only batch by the number of records. By removing FileRecordFormat, we relay the responsibility of implementing the batching to the format developer; they need to use BulkFormat and find a better way than batch by number of records.

## Brief change log

- marked FileRecordFormat and FileRecordFormatAdapter as @Deprecated.
- refactored the relevant usages.
- updated javadoc.

## Verifying this change

This change only marked some classes as deprecated and update relevant comments without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)

### CODE REVIEW

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/FileSource.java [hunk 1]
```diff
@@ -181,12 +178,15 @@ private FileSource(
     }
 
     /**
-     * Builds a new {@code FileSource} using a {@link FileRecordFormat} to read record-by-record
+     * This method is Deprecated, use {@link #forRecordStreamFormat(StreamFormat, Path...)} instead.
```
Comment (slinkydeveloper) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/FileSource.java:181]: Please use the javadoc `@deprecated` tag

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java [hunk 2]
```diff
@@ -37,7 +37,12 @@
 import static org.apache.flink.util.Preconditions.checkArgument;
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
-/** The FormatReaderAdapter turns a {@link FileRecordFormat} into a {@link BulkFormat}. */
+/**
+ * This interface is Deprecated, use {@link StreamFormatAdapter} instead.
```
Comment (slinkydeveloper) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java:41]: Same

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java [hunk 3]
```diff
@@ -34,7 +34,14 @@
 import java.io.Serializable;
 
 /**
- * A reader format that reads individual records from a file.
+ * This interface is Deprecated, use {@link StreamFormat} instead. The main motivation for removing
```
Comment (slinkydeveloper) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java:37]: Same

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java [hunk 4]
```diff
@@ -34,7 +34,14 @@
...
+ * it is the inherent design flaw in the batching of FileRecordFormat: StreamFormat can guarantee
+ * that only a certain amount of memory is being used (unless a single record exceeds that already),
+ * but FileRecordFormat can only batch by the number of records. By removing FileRecordFormat, we
+ * relay the responsibility of implementing the batching to the format developer; they need to use
+ * BulkFormat and find a better way than batch by number of records.
```
Comment (fapaul) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java:42]: I am not sure we really need this docstring here. If someone is interested in the deprecation reason one can open the ticket which is usually linked in the commit message.

**Review by fapaul — COMMENTED** [review]
Thanks for working on this effort. I will definitely, help future format developers to have an easier time when trying to choose the right interface.

I agree with @slinkydeveloper  that the `@deprecated` keyword should be used inside the docstrings.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java (hunk 2 above)
Comment (JingGe) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java:41]: Perfect! Thanks for the info, I will change it.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java (hunk 4 above)
Comment (JingGe) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java:42]: I am open about it and totally fine to remove it. @AHeise WDYT? I thought this case is a little different. Putting this info in Javadoc will give developers the hint while using StreamFormat or BulkFormat as replacement.

Comment (AHeise) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/FileRecordFormat.java:42]: I like to give more information to the developers. However, either rephrase it to match JavaDoc style or just link the ticket.

File: flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapterTest.java [hunk 5]
```diff
@@ -37,6 +37,7 @@
 
 /** Unit and behavior tests for the {@link FileRecordFormatAdapter}. */
 @SuppressWarnings("serial")
+@Deprecated
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapterTest.java:40]: I agree

**Comment by JingGe** [comment]
fixed all issues, @AHeise would you please help merging the PR, thanks.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java [hunk 6]
```diff
@@ -37,7 +37,12 @@
 import static org.apache.flink.util.Preconditions.checkArgument;
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
-/** The FormatReaderAdapter turns a {@link FileRecordFormat} into a {@link BulkFormat}. */
+/**
+ * The FormatReaderAdapter turns a {@link FileRecordFormat} into a {@link BulkFormat}.
+ *
+ * @deprecated Please use {@link StreamFormatAdapter} instead.
```
Comment (fapaul) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java:43]: Nit: I guess this hint is not entirely correct because users cannot use the `StreamFormatAdapter` without migrating to `StreamFormat`. In general, I am not sure whether it is strictly necessary to deprecate internal classes.

**Review by fapaul — APPROVED** [review]
Thanks for the changes the code looks good to me now, only one nitty suggestion.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java (hunk 6 above)
Comment (AHeise) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/FileRecordFormatAdapter.java:43]: I'd go for internal classes as well so that we will not forget to remove them when we remove the public class. Ofc, the link to `StreamFormatAdapter` is debatable but I also don't see that it's hurting anyone.

**Comment by AHeise** [comment]
@flinkbot run azure

**Review by AHeise — APPROVED** [review]
LGTM. Thanks!
